File: properlty/src/test/java/com/ufoscout/properlty/ProperltyTest.java
/*******************************************************************************
 * Copyright 2017 Francesco Cina'
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package com.ufoscout.properlty;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.ufoscout.properlty.reader.ProgrammaticPropertiesReader;
import com.ufoscout.properlty.reader.Properties;
import com.ufoscout.properlty.reader.Reader;
import com.ufoscout.properlty.reader.decorator.ReplacerDecoratorReader;
import org.junit.Test;

import com.ufoscout.properlty.reader.PropertyValue;

import static org.junit.Assert.*;

public class ProperltyTest extends ProperltyBaseTest {

	@Test
	public void shouldReturnEmptyOptionalString() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.get("key.three").isPresent());
	}

	@Test
	public void shouldReturnOptionalString() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertEquals("value.one", prop.get("key.one").get());
		assertEquals("value.two", prop.get("key.two").get());
		assertFalse(prop.get("key.ONE").isPresent());
	}

	@Test
	public void shouldReturnCaseInsensitive() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("KEy.tWo", "value.two");

		final boolean caseSensitive = false;
		final Properlty prop = buildProperlty(properties, caseSensitive);

		assertEquals("value.one", prop.get("key.one").get());
		assertEquals("value.two", prop.get("key.two").get());
		assertEquals("value.one", prop.get("key.ONE").get());
		assertEquals("value.two", prop.get("KEY.TWo").get());
	}

	@Test
	public void shouldReturnDefaultString() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");

		final Properlty prop = buildProperlty(properties);

		assertEquals("value.one", prop.get("key.one", "default"));
		assertEquals("default", prop.get("key.two", "default"));
	}

	@Test
	public void shouldReturnEmptyOptionalInteger() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.getInt("key.three").isPresent());
	}


	@Test
	public void shouldReturnOptionalInteger() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "123");
		properties.put("key.two", "1000000");

		final Properlty prop = buildProperlty(properties);

		assertEquals(123, prop.getInt("key.one").get().intValue());
		assertEquals(1000000, prop.getInt("key.two").get().intValue());
	}

	@Test
	public void shouldReturnDefaultInteger() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "1");

		final Properlty prop = buildProperlty(properties);

		assertEquals(1, prop.getInt("key.one", 10));
		assertEquals(10, prop.getInt("key.two", 10));

	}

	@Test(expected=NumberFormatException.class)
	public void shouldThrowExceptionParsingWrongInteger() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "not a number");

		final Properlty prop = buildProperlty(properties);

		prop.getInt("key.one");

	}

	@Test
	public void shouldReturnEmptyOptionalBoolean() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.getBoolean("key.three").isPresent());
	}


	@Test
	public void shouldReturnOptionalBoolean() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "True");
		properties.put("key.two", "false");

		final Properlty prop = buildProperlty(properties);

		assertTrue(prop.getBoolean("key.one").get());
		assertFalse(prop.getBoolean("key.two").get());
	}

	@Test
	public void shouldReturnDefaultBoolean() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "true");

		final Properlty prop = buildProperlty(properties);

		assertTrue(prop.getBoolean("key.one", false));
		assertFalse(prop.getBoolean("key.two", false));
		assertTrue(prop.getBoolean("key.two", true));

	}

	@Test(expected=RuntimeException.class)
	public void shouldThrowExceptionParsingWrongBoolean() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "not a bool");

		final Properlty prop = buildProperlty(properties);

		prop.getBoolean("key.one");

	}

	@Test
	public void shouldReturnEmptyOptionalDouble() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.getDouble("key.three").isPresent());
	}


	@Test
	public void shouldReturnOptionalDouble() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "123");
		properties.put("key.two", "1000000");

		final Properlty prop = buildProperlty(properties);

		assertEquals(123d, prop.getDouble("key.one").get().doubleValue(), 0.1d);
		assertEquals(1000000d, prop.getDouble("key.two").get().doubleValue(), 0.1d);
	}

	@Test
	public void shouldReturnDefaultDouble() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "1");

		final Properlty prop = buildProperlty(properties);

		assertEquals(1d, prop.getDouble("key.one", 1.1111d), 0.1d);
		assertEquals(10d, prop.getDouble("key.two", 10d), 0.1d);

	}

	@Test(expected=NumberFormatException.class)
	public void shouldThrowExceptionParsingWrongDouble() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "not a number");

		final Properlty prop = buildProperlty(properties);

		prop.getDouble("key.one");

	}

	@Test
	public void shouldReturnEmptyOptionalFloat() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.getFloat("key.three").isPresent());
	}


	@Test
	public void shouldReturnOptionalFloat() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "123");
		properties.put("key.two", "1000000");

		final Properlty prop = buildProperlty(properties);

		assertEquals(123f, prop.getFloat("key.one").get().floatValue(), 0.1f);
		assertEquals(1000000f, prop.getFloat("key.two").get().floatValue(), 0.1f);
	}

	@Test
	public void shouldReturnDefaultFloat() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "1");

		final Properlty prop = buildProperlty(properties);

		assertEquals(1f, prop.getFloat("key.one", 10), 0.1f);
		assertEquals(10f, prop.getFloat("key.two", 10), 0.1f);

	}

	@Test(expected=NumberFormatException.class)
	public void shouldThrowExceptionParsingWrongFloat() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "not a number");

		final Properlty prop = buildProperlty(properties);

		prop.getFloat("key.one");

	}

	@Test
	public void shouldReturnEmptyOptionalLong() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.getLong("key.three").isPresent());
	}


	@Test
	public void shouldReturnOptionalLong() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "123");
		properties.put("key.two", "1000000");

		final Properlty prop = buildProperlty(properties);

		assertEquals(123l, prop.getLong("key.one").get().longValue());
		assertEquals(1000000l, prop.getLong("key.two").get().longValue());
	}

	@Test
	public void shouldReturnDefaultLong() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "1");

		final Properlty prop = buildProperlty(properties);

		assertEquals(1l, prop.getLong("key.one", 10l));
		assertEquals(10l, prop.getLong("key.two", 10l));

	}

	@Test
	public void shouldReturnOptionalBigDecimal() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "123");
		properties.put("key.two", "1000000");

		final Properlty prop = buildProperlty(properties);

		assertEquals(123, prop.getBigDecimal("key.one").get().intValue());
		assertEquals(1000000, prop.getBigDecimal("key.two").get().intValue());
	}

	@Test
	public void shouldReturnDefaultBigDecimal() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "1");

		final Properlty prop = buildProperlty(properties);

		assertEquals(1, prop.getBigDecimal("key.one", new BigDecimal(10)).intValue());
		assertEquals(10, prop.getBigDecimal("key.two", new BigDecimal(10)).intValue());

	}

	@Test
	public void shouldReturnOptionalBigInteger() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "123");
		properties.put("key.two", "1000000");

		final Properlty prop = buildProperlty(properties);

		assertEquals(123, prop.getBigInteger("key.one").get().intValue());
		assertEquals(1000000, prop.getBigInteger("key.two").get().intValue());
	}

	@Test
	public void shouldReturnDefaultBigInteger() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "1");

		final Properlty prop = buildProperlty(properties);

		assertEquals(1, prop.getBigInteger("key.one", new BigInteger("10")).intValue());
		assertEquals(10, prop.getBigInteger("key.two", new BigInteger("10")).intValue());

	}

	@Test(expected=NumberFormatException.class)
	public void shouldThrowExceptionParsingWrongLong() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "not a number");

		final Properlty prop = buildProperlty(properties);

		prop.getLong("key.one");

	}

	@Test
	public void shouldReturnEmptyOptionalEnum() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "value.one");
		properties.put("key.two", "value.two");

		final Properlty prop = buildProperlty(properties);

		assertFalse(prop.getEnum("key.three", NeedSomebodyToLove.class).isPresent());
	}


	@Test
	public void shouldReturnOptionalEnum() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "ME");
		properties.put("key.two", "THEM");

		final Properlty prop = buildProperlty(properties);

		assertEquals(NeedSomebodyToLove.ME, prop.getEnum("key.one", NeedSomebodyToLove.class).get());
		assertEquals(NeedSomebodyToLove.THEM, prop.getEnum("key.two", NeedSomebodyToLove.class).get());
	}

	@Test
	public void shouldReturnDefaultEnum() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "ME");

		final Properlty prop = buildProperlty(properties);

		assertEquals(NeedSomebodyToLove.ME, prop.getEnum("key.one", NeedSomebodyToLove.THEM));
		assertEquals(NeedSomebodyToLove.THEM, prop.getEnum("key.two", NeedSomebodyToLove.THEM));

	}

	@Test(expected=IllegalArgumentException.class)
	public void shouldThrowExceptionParsingWrongEnumLong() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "not an enum");

		final Properlty prop = buildProperlty(properties);

		prop.getEnum("key.one", NeedSomebodyToLove.class);

	}

	@Test
	public void shouldReturntheKeyApplyingTheMapFunction() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "111");

		final Properlty prop = buildProperlty(properties);

		assertEquals(111, prop.get("key.one", Integer::valueOf).get().intValue());
		assertEquals(222, prop.get("key.two", 222, Integer::valueOf).intValue());
		assertFalse(prop.get("key.three", Integer::valueOf).isPresent());
	}

	@Test
	public void shouldReturnValueToArray() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "111,AAAAA,BBB");

		final Properlty prop = buildProperlty(properties);

		final String[] values = prop.getArray("key.one");
		assertEquals(3, values.length);
		assertEquals("111", values[0]);
		assertEquals("AAAAA", values[1]);
		assertEquals("BBB", values[2]);
		assertEquals(0, prop.getArray("key.three").length);
	}

	@Test
	public void shouldReturnValueToList() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "111,AAAAA,BBB");

		final Properlty prop = buildProperlty(properties);

		final List<String> values = prop.getList("key.one");
		assertEquals(3, values.size());
		assertEquals("111", values.get(0));
		assertEquals("AAAAA", values.get(1));
		assertEquals("BBB", values.get(2));
		assertEquals(0, prop.getList("key.three").size());
	}

	@Test
	public void shouldReturnValueToListOfCustomObjects() {
		final Map<String, String> properties = new HashMap<>();

		properties.put("key.one", "111,222,333");

		final Properlty prop = buildProperlty(properties);

		final List<Integer> values = prop.getList("key.one", Integer::valueOf);
		assertEquals(3, values.size());
		assertEquals(111, values.get(0).intValue());
		assertEquals(222, values.get(1).intValue());
		assertEquals(333, values.get(2).intValue());
		assertEquals(0, prop.getList("key.three", Integer::valueOf).size());
	}

	@Test
	public void shouldMatchPlaceholdersNotSensitiveCase() {
		final Map<String, String> properties = new HashMap<>();
		properties.put("key.ONE", "${value1:defaultValue1}");
		properties.put("keY.TWO", "${KEY.one:defaultValue2}");
		properties.put("key.three", "${KEY.one:defaultValue3}");

		final boolean caseSensitive = false;
		final Properlty prop = buildProperlty(properties, caseSensitive);

		assertEquals("defaultValue1", prop.get("key.one").get());
		assertEquals("defaultValue1", prop.get("key.two").get());
		assertEquals("defaultValue1", prop.get("key.THREE").get());

	}

	@Test
	public void shouldMatchPlaceholdersSensitiveCase() {
		final Map<String, String> properties = new HashMap<>();
		properties.put("key.ONE", "${value1:defaultValue1}");
		properties.put("keY.TWO", "${KEY.one:defaultValue2}");
		properties.put("key.three", "${KEY.one:defaultValue3}");

		final boolean caseSensitive = true;
		final Properlty prop = buildProperlty(properties, caseSensitive);

		assertEquals("defaultValue1", prop.get("key.ONE").get());
		assertEquals("defaultValue2", prop.get("keY.TWO").get());
		assertEquals("defaultValue3", prop.get("key.three").get());

	}

	private Properlty buildProperlty(Map<String, String> properties) {
		return buildProperlty(properties, true);
	}

	private Properlty buildProperlty(Map<String, String> properties, boolean caseSensitive) {
		ProperltyBuilder builder = Properlty.builder();
		properties.forEach((key, value) -> {
			builder.add(Properties.add(key, value));
		});
		return builder.caseSensitive(caseSensitive).build();
	}

}

File: properlty-kotlin/src/test/java/com/ufoscout/properlty/ProperltyTest.kt
/*******************************************************************************
 * Copyright 2017 Francesco Cina'

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ufoscout.properlty

import com.ufoscout.properlty.reader.Properties
import com.ufoscout.properlty.reader.PropertyValue
import org.junit.Assert.*
import org.junit.Test
import java.lang.RuntimeException
import java.math.BigDecimal
import java.math.BigInteger
import java.util.*
import java.util.stream.Collectors

class ProperltyTest : ProperltyBaseTest() {

    @Test
    fun shouldReturnEmptyOptionalString() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop["key.three"])
    }

    @Test
    fun shouldReturnOptionalString() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertEquals("value.one", prop.get("key.one"))
        assertEquals("value.two", prop["key.two"])
        assertNull(prop["key.ONE"])
    }

    @Test
    fun shouldReturnCaseInsensitive() {
        val properties = HashMap<String, String>()

        properties["key.one"] = "value.one"
        properties["KEy.tWo"] = "value.two"

        val caseSensitive = false
        val prop = buildProperlty(properties, caseSensitive)

        assertEquals("value.one", prop["key.one"])
        assertEquals("value.two", prop["key.two"])
        assertEquals("value.one", prop["key.ONE"])
        assertEquals("value.two", prop["KEY.TWo"])
    }

    @Test
    fun shouldReturnDefaultString() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")

        val prop = buildProperlty(properties)

        assertEquals("value.one", prop["key.one", "default"])
        assertEquals("default", prop["key.two", "default"])
    }

    @Test
    fun shouldReturnEmptyOptionalInteger() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop.getInt("key.three"))
    }


    @Test
    fun shouldReturnOptionalInteger() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "123")
        properties.put("key.two", "1000000")

        val prop = buildProperlty(properties)

        assertEquals(123, prop.getInt("key.one"))
        assertEquals(1000000, prop.getInt("key.two"))
    }

    @Test
    fun shouldReturnDefaultInteger() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "1")

        val prop = buildProperlty(properties)

        assertEquals(1, prop.getInt("key.one", 10))
        assertEquals(10, prop.getInt("key.two", 10))

    }

    @Test(expected = NumberFormatException::class)
    fun shouldThrowExceptionParsingWrongInteger() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "not a number")

        val prop = buildProperlty(properties)

        prop.getInt("key.one")

    }

    @Test
    fun shouldReturnEmptyOptionalBoolean() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop.getBoolean("key.three"))
    }


    @Test
    fun shouldReturnOptionalBoolean() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "true")
        properties.put("key.two", "false")

        val prop = buildProperlty(properties)

        assertTrue(prop.getBoolean("key.one")!!)
        assertFalse(prop.getBoolean("key.two")!!)
    }

    @Test
    fun shouldReturnDefaultBoolean() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "True")

        val prop = buildProperlty(properties)

        assertTrue(prop.getBoolean("key.one", false))
        assertFalse(prop.getBoolean("key.two", false))
        assertTrue(prop.getBoolean("key.two", true))

    }

    @Test(expected = RuntimeException::class)
    fun shouldThrowExceptionParsingWrongBoolean() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "not a boolean")

        val prop = buildProperlty(properties)

        prop.getDouble("key.one")

    }

    @Test
    fun shouldReturnEmptyOptionalDouble() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop.getDouble("key.three"))
    }


    @Test
    fun shouldReturnOptionalDouble() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "123")
        properties.put("key.two", "1000000")

        val prop = buildProperlty(properties)

        assertEquals(123.0, prop.getDouble("key.one")!!, 0.1)
        assertEquals(1000000.0, prop.getDouble("key.two")!!, 0.1)
    }

    @Test
    fun shouldReturnDefaultDouble() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "1")

        val prop = buildProperlty(properties)

        assertEquals(1.0, prop.getDouble("key.one", 1.1111), 0.1)
        assertEquals(10.0, prop.getDouble("key.two", 10.0), 0.1)

    }

    @Test(expected = NumberFormatException::class)
    fun shouldThrowExceptionParsingWrongDouble() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "not a number")

        val prop = buildProperlty(properties)

        prop.getDouble("key.one")

    }

    @Test
    fun shouldReturnEmptyOptionalFloat() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop.getFloat("key.three"))
    }


    @Test
    fun shouldReturnOptionalFloat() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "123")
        properties.put("key.two", "1000000")

        val prop = buildProperlty(properties)

        assertEquals(123f, prop.getFloat("key.one")!!, 0.1f)
        assertEquals(1000000f, prop.getFloat("key.two")!!, 0.1f)
    }

    @Test
    fun shouldReturnDefaultFloat() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "1")

        val prop = buildProperlty(properties)

        assertEquals(1f, prop.getFloat("key.one", 10f), 0.1f)
        assertEquals(10f, prop.getFloat("key.two", 10f), 0.1f)

    }

    @Test(expected = NumberFormatException::class)
    fun shouldThrowExceptionParsingWrongFloat() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "not a number")

        val prop = buildProperlty(properties)

        prop.getFloat("key.one")

    }

    @Test
    fun shouldReturnEmptyOptionalLong() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop.getLong("key.three"))
    }


    @Test
    fun shouldReturnOptionalLong() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "123")
        properties.put("key.two", "1000000")

        val prop = buildProperlty(properties)

        assertEquals(123L, prop.getLong("key.one"))
        assertEquals(1000000L, prop.getLong("key.two"))
    }

    @Test
    fun shouldReturnDefaultLong() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "1")

        val prop = buildProperlty(properties)

        assertEquals(1L, prop.getLong("key.one", 10L))
        assertEquals(10L, prop.getLong("key.two", 10L))

    }

    @Test
    fun shouldReturnOptionalBigDecimal() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "123")
        properties.put("key.two", "1000000")

        val prop = buildProperlty(properties)

        assertEquals(123L, prop.getBigDecimal("key.one")?.toLong())
        assertEquals(1000000L, prop.getBigDecimal("key.two")?.toLong())
    }

    @Test
    fun shouldReturnDefaultBigDecimal() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "1")

        val prop = buildProperlty(properties)

        assertEquals(1L, prop.getBigDecimal("key.one", BigDecimal("10")).toLong())
        assertEquals(10L, prop.getBigDecimal("key.two", BigDecimal(10L)).toLong())

    }

    @Test
    fun shouldReturnOptionalBigInteger() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "123")
        properties.put("key.two", "1000000")

        val prop = buildProperlty(properties)

        assertEquals(123L, prop.getBigInteger("key.one")?.toLong())
        assertEquals(1000000L, prop.getBigInteger("key.two")?.toLong())
    }

    @Test
    fun shouldReturnDefaultBigInteger() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "1")

        val prop = buildProperlty(properties)

        assertEquals(1L, prop.getBigInteger("key.one", BigInteger.valueOf(10L)).toLong())
        assertEquals(10L, prop.getBigInteger("key.two", BigInteger.valueOf(10L)).toLong())

    }

    @Test(expected = NumberFormatException::class)
    fun shouldThrowExceptionParsingWrongLong() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "not a number")

        val prop = buildProperlty(properties)

        prop.getLong("key.one")

    }

    @Test
    fun shouldReturnEmptyOptionalEnum() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "value.one")
        properties.put("key.two", "value.two")

        val prop = buildProperlty(properties)

        assertNull(prop.getEnum<NeedSomebodyToLove>("key.three"))
    }


    @Test
    fun shouldReturnOptionalEnum() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "ME")
        properties.put("key.two", "THEM")

        val prop = buildProperlty(properties)

        assertEquals(NeedSomebodyToLove.ME, prop.getEnum<NeedSomebodyToLove>("key.one"))
        assertEquals(NeedSomebodyToLove.THEM, prop.getEnum<NeedSomebodyToLove>("key.two"))
    }

    @Test
    fun shouldReturnDefaultEnum() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "ME")

        val prop = buildProperlty(properties)

        assertEquals(NeedSomebodyToLove.ME, prop.getEnum("key.one", NeedSomebodyToLove.THEM))
        assertEquals(NeedSomebodyToLove.THEM, prop.getEnum("key.two", NeedSomebodyToLove.THEM))

    }

    @Test(expected = IllegalArgumentException::class)
    fun shouldThrowExceptionParsingWrongEnumLong() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "not an enum")

        val prop = buildProperlty(properties)

        prop.getEnum<NeedSomebodyToLove>("key.one")

    }

    @Test
    fun shouldReturntheKeyApplyingTheMapFunction() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "111")

        val prop = buildProperlty(properties)

        assertEquals(111, prop.get<Int>("key.one", { Integer.valueOf(it) }))
        assertEquals(222, prop["key.two", 222, { Integer.valueOf(it) }])
        assertNull(prop.get<Int>("key.three", { Integer.valueOf(it) }))
    }

    @Test
    fun shouldReturnValueToArray() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "111,AAAAA,BBB")

        val prop = buildProperlty(properties)

        val values = prop.getArray("key.one")
        assertEquals(3, values.size)
        assertEquals("111", values[0])
        assertEquals("AAAAA", values[1])
        assertEquals("BBB", values[2])
        assertEquals(0, prop.getArray("key.three").size)
    }

    @Test
    fun shouldReturnValueToList() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "111,AAAAA,BBB")

        val prop = buildProperlty(properties)

        val values = prop.getList("key.one")
        assertEquals(3, values.size)
        assertEquals("111", values[0])
        assertEquals("AAAAA", values[1])
        assertEquals("BBB", values[2])
        assertEquals(0, prop.getList("key.three").size)
    }

    @Test
    fun shouldReturnValueToListOfCustomObjects() {
        val properties = HashMap<String, String>()

        properties.put("key.one", "111,222,333")

        val prop = buildProperlty(properties)

        val values = prop.getList<Int>(key = "key.one"){Integer.valueOf(it)}
        assertEquals(3, values.size)
        assertEquals(111, values[0])
        assertEquals(222, values[1])
        assertEquals(333, values[2])
        assertEquals(0, prop.getList<Int>("key.three", { Integer.valueOf(it) }).size)
    }

    @Test
    fun shouldMatchPlaceholdersNotSensitiveCase() {
        val properties = HashMap<String, String>()
        properties["key.ONE"] = "\${value1:defaultValue1}"
        properties["keY.TWO"] = "\${KEY.one:defaultValue2}"
        properties["key.three"] = "\${KEY.one:defaultValue3}"

        val caseSensitive = false
        val prop = buildProperlty(properties, caseSensitive)

        assertEquals("defaultValue1", prop["key.one"])
        assertEquals("defaultValue1", prop["key.two"])
        assertEquals("defaultValue1", prop["key.THREE"])

    }

    @Test
    fun shouldMatchPlaceholdersSensitiveCase() {
        val properties = HashMap<String, String>()
        properties["key.ONE"] = "\${value1:defaultValue1}"
        properties["keY.TWO"] = "\${KEY.one:defaultValue2}"
        properties["key.three"] = "\${KEY.one:defaultValue3}"

        val caseSensitive = true
        val prop = buildProperlty(properties, caseSensitive)

        assertEquals("defaultValue1", prop["key.ONE"])
        assertEquals("defaultValue2", prop["keY.TWO"])
        assertEquals("defaultValue3", prop["key.three"])

    }

    private fun buildProperlty(properties: Map<String, String>, caseSensitive: Boolean = true): Properlty {
        val builder = Properlty.builder().caseSensitive(caseSensitive)
        properties.forEach { key, value -> builder.add(Properties.add(key, value)) }
        return builder.build()
    }

}
